x-pack/plugins/synthetics/server/synthetics_service/project_monitor/normalizers/index.ts

```diff
@@ -0,0 +1,158 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { omit } from 'lodash';
+import {
+  BrowserFields,
+  ConfigKey,
+  DataStream,
+  PrivateLocation,
+  Locations,
+  ProjectMonitor,
+} from '../../../../common/runtime_types';
+import { DEFAULT_FIELDS } from '../../../../common/constants/monitor_defaults';
+import { getNormalizeBrowserFields } from './browser_monitor';
+import { getNormalizeICMPFields } from './icmp_monitor';
+import { getNormalizeTCPFields } from './tcp_monitor';
+import { getNormalizeHTTPFields } from './http_monitor';
+export { getCustomHeartbeatId } from './common_fields';
+
+export interface NormalizedProjectProps {
+  locations: Locations;
+  privateLocations: PrivateLocation[];
+  monitor: ProjectMonitor;
+  projectId: string;
+  namespace: string;
+}
+
+export const normalizeProjectMonitor = (props: NormalizedProjectProps) => {
+  const { monitor } = props;
+  const type = monitor.type || DataStream.BROWSER;
+
+  switch (type) {
+    case DataStream.BROWSER:
+      return getNormalizeBrowserFields(props);
+
+    case DataStream.HTTP:
+      return getNormalizeHTTPFields(props);
+
+    case DataStream.TCP:
+      return getNormalizeTCPFields(props);
+
+    case DataStream.ICMP:
+      return getNormalizeICMPFields(props);
+    default:
+      throw new Error(`Unsupported monitor type ${monitor.type}`);
+  }
+};
+
+export const normalizeProjectMonitors = ({
+  locations = [],
+  privateLocations = [],
+  monitors = [],
+  projectId,
+  namespace,
+}: {
+  locations: Locations;
+  privateLocations: PrivateLocation[];
+  monitors: ProjectMonitor[];
+  projectId: string;
+  namespace: string;
+}) => {
+  return monitors.map((monitor) => {
+    return normalizeProjectMonitor({ monitor, locations, privateLocations, projectId, namespace });
+  });
+};
+
+export const getMonitorLocations = ({
+  privateLocations,
+  publicLocations,
+  monitor,
+}: {
+  monitor: ProjectMonitor;
+  privateLocations: PrivateLocation[];
+  publicLocations: Locations;
+}) => {
+  const publicLocs =
+    monitor.locations?.map((id) => {
+      return publicLocations.find((location) => location.id === id);
+    }) || [];
+  const privateLocs =
+    monitor.privateLocations?.map((locationName) => {
+      return privateLocations.find(
+        (location) =>
+          location.label.toLowerCase() === locationName.toLowerCase() ||
+          location.id.toLowerCase() === locationName.toLowerCase()
+      );
+    }) || [];
+
+  return [...publicLocs, ...privateLocs].filter(
+    (location) => location !== undefined
+  ) as BrowserFields[ConfigKey.LOCATIONS];
+};
+
+export const normalizeYamlConfig = (monitor: NormalizedProjectProps['monitor']) => {
+  const defaultFields = DEFAULT_FIELDS[monitor.type as DataStream];
+
+  const flattenAndFormatObject = (obj: Record<string, unknown>, prefix = '', keys: string[]) =>
```

Comment:
i thought agent would flatten fields?

Reply:
Discussed with @vigneshshanmugam and we decided to do it on the Kibana side. Here's why:

We identified that the agent, flattening indiscriminately, would overflatten keys. 
For example, consider
```
      "check.request": {
        "method": "POST",
        "headers": {
          "Content-Type": "application/x-www-form-urlencoded"
        }
      },
```
With agent, who doesn't have context of the keys supported by Kibana, the flattening logic would flatten these keys to `check.request.method` and `check.request.headers.Content-Type`. 

We only support `check.request.headers`, however, and don't further flatten the user-provided json. 

Kibana has knowledge of the keys it supports. That's why we can do logic [here](https://github.com/elastic/kibana/pull/140713/files#diff-054c9e026a0c7f90da34f2d44ca3537f395a4a7acb1ea0c4dae7d16d93343344R113) that says stop flattening early once we spot a supported key.